### PR TITLE
Add ability to dynamically set nested schema type

### PR DESCRIPTION
**Pull request by OliverJAsh**
# Problem

Reinstates https://github.com/paularmstrong/normalizr/pull/159, which [was unfortunately lost when v3 was released](https://github.com/paularmstrong/normalizr/issues/130#issuecomment-285012859).

Fixes https://github.com/paularmstrong/normalizr/issues/413

Related: https://github.com/paularmstrong/normalizr/issues/130

I will update documentation once the code side of things has been approved.

/cc @paularmstrong @jgutta @fxck @joewestcott

# TODO

- [x] Add & update tests
- [x] Ensure CI is passing (lint, tests, flow)
- [ ] Update relevant documentation

### CODE REVIEW

**Comment by OliverJAsh** [comment]
Hi @paularmstrong. Is there anyone who maintains normalizr that might be available to review this?

**Comment by OliverJAsh** [comment]
Hi @paularmstrong @gaearon. Sorry to pester but this is blocking us at Unsplash from upgrading from v2 to v3. Would you mind reviewing, or do you know anyone who can? Thanks 😄

**Comment by OliverJAsh** [comment]
Ping @paularmstrong

File: typescript-tests/object.ts [hunk 1]
```diff
@@ -1,12 +1,22 @@
...
-};
+type Response = {
+  users: Array<{ id: string }>
+}
+const data: Response = { users: [ { id: 'foo' } ] };
 const user = new schema.Entity('users');
 
-const responseSchema = new schema.Object({ users: new schema.Array(user) });
-const normalizedData = normalize(data, responseSchema);
+{
```
Comment (paularmstrong) [inline typescript-tests/object.ts]: What do these added brackets do?

Comment (OliverJAsh) [reply typescript-tests/object.ts]: They provide unique block scope for each of the tests, so I can redeclare variables using the same name.

**Comment by BrentFarese** [comment]
Coming across this a while after it was posted/implemented, but has anyone noticed that this approach doesn't seem to work when calling `denormalize`?  We had the approach working successfully for `normalize`, but `denormalize` fails always with `schema.denormalize is not a function`.

**Pull request merged**